src/components/select/AsyncSelect.tsx

```diff
@@ -0,0 +1,138 @@
+/*
+ * Echoes React
+ * Copyright (C) 2023-2024 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+import {
+  Select as MantineSelect,
+  SelectProps as MantineSelectProps,
+  SelectItem,
+} from '@mantine/core';
+import { forwardRef, useCallback, useRef } from 'react';
+import { PropsWithLabels } from '~types/utils';
+import { Spinner } from '..';
+import { SelectHighlight } from './Select';
+
+interface Props {
+  className?: string;
+  data: ReadonlyArray<SelectItem>;
+  hasError?: boolean;
+  highlight?: SelectHighlight;
+  id?: string;
+  isDisabled?: boolean;
+  isLoading?: boolean;
+  isNotClearable?: boolean;
+  isRequired?: boolean;
+  labelError?: MantineSelectProps['error'];
+  labelNotFound?: MantineSelectProps['nothingFound'];
+  optionComponent?: MantineSelectProps['itemComponent'];
+  onChange?: MantineSelectProps['onChange'];
+  onSearch?: MantineSelectProps['onSearchChange'];
+  value?: MantineSelectProps['value'];
+}
+
+export const AsyncSelect = forwardRef<HTMLInputElement, PropsWithLabels<Props>>((props, ref) => {
+  const {
+    ariaLabel,
+    data,
+    hasError,
+    helpText,
+    highlight,
+    isNotClearable,
+    isDisabled,
+    isLoading,
+    isRequired,
+    label,
+    labelError,
+    labelNotFound,
+    optionComponent,
+    onChange,
+    onSearch,
+    value,
+  } = props;
+
+  const previousQuery = useRef<string>();
+  const valueRef = useRef<string | null>();
+
+  const handleSearch = useCallback(
+    (query: string) => {
+      // Prevent search from being triggered when we select a value
+      const currentValue = getItemLabel(
+        data.find((item) => getItemValue(item) === valueRef.current),
+      );
+
+      if (onSearch && query !== previousQuery.current && query !== currentValue) {
+        previousQuery.current = query;
+        onSearch(query);
+      }
+    },
+    [data, onSearch],
+  );
+
+  const handleChange = useCallback(
+    (value: string | null) => {
+      valueRef.current = value;
+
+      if (onChange) {
+        onChange(value);
+      }
+    },
+    [onChange],
+  );
+
+  return (
+    <MantineSelect
+      allowDeselect
```

Comment:
we should activate this only we have `!isNotClearable && !isRequired` right? and we should also add it in the normal select I guess?

Reply:
Good point, but only `!isRequired` I think.

Reply:
If it should not be clearable then we shouldn't be able to unselect no ?